### PR TITLE
Use objects, not strings, for assertions

### PR DESCRIPTION
**Pull request by muglug**
It's a big addition, but this would make it slightly easier to track where assertions were being used in the codebase, and removes a lot of checks for `$assertion[0] === '!'` etc.

### CODE REVIEW

**Comment by orklah** [comment]
This looks really good!

**Comment by orklah** [comment]
Hi @muglug ! Do you think you'll finish this soon? I'd really like having this in the alpha for Psalm 5!

**Comment by muglug** [comment]
Yup, should have unit tests passing today

**Comment by orklah** [comment]
Sorry, my last merge on master broke CI. I'll fix that

File: src/Psalm/Internal/Type/AssertionReconciler.php [hunk 1]
```diff
@@ -1257,21 +1179,15 @@ private static function handleLiteralEqualityWithString(
                     return $existing_var_type;
                 }
 
-                if ($scalar_type === 'class-string'
-                    || $scalar_type === 'interface-string'
-                    || $scalar_type === 'trait-string'
-                ) {
+                if ($assertion_type instanceof TLiteralClassString) {
```
Comment (orklah) [inline src/Psalm/Internal/Type/AssertionReconciler.php:1182]: this condition can be dropped (there are a few like that in the file)

**Comment by orklah** [comment]
Could you add some details on UPDGRADE.md to explain that the old Assertion was renamed and other BC breaks introduced?

File: src/Psalm/Type/Atomic/TClassString.php [hunk 2]
```diff
@@ -51,7 +58,9 @@ public function __toString(): string
 
     public function getId(bool $nested = false): string
     {
-        return $this->getKey();
+        return ($this->is_loaded ? 'loaded-' : '')
+            . ($this->is_interface ? 'interface' : 'class')
+            . '-string' . ($this->as === 'object' ? '' : '<' . $this->as_type . '>');
     }
 
     public function getAssertionString(bool $exact = false): string
```
Comment (orklah) [inline src/Psalm/Type/Atomic/TClassString.php:66]: every call to getAssertionString is now with $exact to true. Can we drop the param?

**Comment by orklah** [comment]
every call to getAssertionString is now with $exact to true. Can we drop the param?

File: src/Psalm/Internal/Type/AssertionReconciler.php (hunk 1 above)
Comment (muglug) [reply src/Psalm/Internal/Type/AssertionReconciler.php:1182]: Why can it be dropped?

File: src/Psalm/Type/Atomic/TCallableObject.php [hunk 3]
```diff
@@ -34,7 +34,9 @@ public function canBeFullyExpressedInPhp(int $analysis_php_version_id): bool
         return false;
     }
 
-    public function getAssertionString(bool $exact = false): string
+    public function getAssertionString()
+
+    /** @psalm-mutation-free */: string
```
Comment (orklah) [inline src/Psalm/Type/Atomic/TCallableObject.php:39]: this feels weird

File: src/Psalm/Internal/Type/AssertionReconciler.php (hunk 1 above)
Comment (muglug) [reply src/Psalm/Internal/Type/AssertionReconciler.php]: Done!

File: tests/TypeReconciliation/EmptyTest.php [hunk 4]
```diff
@@ -473,19 +473,6 @@ function foo(array $r) {
                     }',
                 'error_message' => 'MixedReturnTypeCoercion'
             ],
-            'secondEmptyTwice' => [
```
Comment (orklah) [inline tests/TypeReconciliation/EmptyTest.php:476]: why did you drop this? Seems like it still works on my side

Comment (muglug) [reply tests/TypeReconciliation/EmptyTest.php:476]: I migrated the test (in the more common falsiness check form) to `RedundantConditionTest.php`

File: src/Psalm/Internal/Type/AssertionReconciler.php (hunk 1 above)
Comment (orklah) [reply src/Psalm/Internal/Type/AssertionReconciler.php]: The one on line 1456 can still be dropped :)

**Comment by orklah** [comment]
This makes so many things looks a lot more clearer! Very cool! Thanks :)

File: src/Psalm/Storage/Assertion/Empty_.php [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    }
+
+    public function isNegation(): bool
+    {
+        return true;
+    }
+
+    public function __toString(): string
+    {
+        return '!non-empty';
```
[the remark this answers is not in the document]
Comment (orklah) [reply src/Psalm/Storage/Assertion/Empty_.php:22]: Somehow it still works? https://psalm.dev/r/bb6d82a6a3 https://psalm.dev/r/90da8dba12

I didn't expect that. What is the system that parses assertion into Assertion objects?

[the remark this answers is not in the document]
Comment (AndrolGenhald) [reply src/Psalm/Storage/Assertion/Empty_.php:22]: @orklah It's the [scanner](https://github.com/vimeo/psalm/blob/2e01e9b7cdefed310ee506abb2bd063ba380f883/src/Psalm/Internal/PhpVisitor/Reflector/FunctionLikeDocblockScanner.php#L594).

[the remark this answers is not in the document]
Comment (orklah) [reply src/Psalm/Storage/Assertion/Empty_.php:22]: Thanks! I never had to tweak here before... Weird, I thought there would have been a direct link between internal assertion names and user assertions ones. But this answer my question! Thanks both!

**Pull request merged**